### PR TITLE
Feature/create event

### PR DESCRIPTION
**Pull request by JavaBaz**

### CODE REVIEW

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+@Data
+@SuperBuilder
+@MappedSuperclass
+@NoArgsConstructor
+
+public class BaseEntity<ID extends Serializable> implements Serializable {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java]: why use identity? why not use other generation types such as sequence?

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.io.Serializable;
+
+
+@Data
+@SuperBuilder
+@MappedSuperclass
+@NoArgsConstructor
+
+public class BaseEntity<ID extends Serializable> implements Serializable {
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java]: why use an entity just to hold the id? We know that all entities have id field. Why not put it in each entity?

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/repository/BaseRepository.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package ir.tehran.jug.projects.eventmanagement.data.base.repository;
+
+import ir.tehran.jug.projects.eventmanagement.data.base.entity.BaseEntity;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.io.Serializable;
+
+public interface BaseRepository<E extends BaseEntity<ID>, ID extends Serializable> extends JpaRepository<E, ID> {
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/repository/BaseRepository.java]: I suggest to move this class to repositories folder

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java (hunk 2 above)
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java]: I suggest to move this class to entities package

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package ir.tehran.jug.projects.eventmanagement.data.base.service;
+
+import ir.tehran.jug.projects.eventmanagement.data.base.entity.BaseEntity;
+
+import java.io.Serializable;
+import java.util.List;
+
+public interface BaseService<E extends BaseEntity<ID>, ID extends Serializable> {
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java]: remove this class and its implementation. The repositories do the same thing

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/Category.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.Set;
+
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@SuperBuilder
+@Entity
+@Table(name = "CATEGORIES")
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/Category.java]: Do some research on naming convention for postgresql tables. take a look at the following link:
https://stackoverflow.com/questions/2878248/postgresql-naming-conventions

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/events/Event.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
+@SuperBuilder
+@Entity
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/events/Event.java:24]: Why not use other types of inheritance?

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/enums/Platform.java [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package ir.tehran.jug.projects.eventmanagement.data.entities.enums;
+
+public enum Platform {
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/enums/Platform.java]: Platform is a little misleading. Can we find a better name?

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/TagRepository.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package ir.tehran.jug.projects.eventmanagement.data.repositories;
+
+import ir.tehran.jug.projects.eventmanagement.data.base.repository.BaseRepository;
+import ir.tehran.jug.projects.eventmanagement.data.entities.Tag;
+
+public interface TagRepository extends BaseRepository<Tag, Long> {
+    Tag findByTitle(String tagTitle);
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/TagRepository.java]: use optional as return type.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/CategoryRepository.java [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package ir.tehran.jug.projects.eventmanagement.data.repositories;
+
+import ir.tehran.jug.projects.eventmanagement.data.base.repository.BaseRepository;
+import ir.tehran.jug.projects.eventmanagement.data.entities.Category;
+
+public interface CategoryRepository extends BaseRepository<Category, Long> {
+    Category findByTitle(String categoryTitle);
```
Comment (hesamghiasi) [inline src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/CategoryRepository.java]: why did you add one method in all repositories? We don't have any use case for them.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/repository/BaseRepository.java (hunk 3 above)
Comment (hryesf) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/repository/BaseRepository.java]: Why we should have BaseRepository? 
Is this class avoid complexity or make complexity?

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java (hunk 4 above)
Comment (hryesf) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java]: I think in three base interfaces that is wrote the BaseEntity is useful and other base interfaces can make complexity in our code.

Comment (hryesf) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java]: Spring data JPA provide abstraction for us

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java (hunk 1 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java]: In this project, batch inserting isn't a requirement, making the retrieval of the next 50 sequences less relevant. Additionally, opting for sequence strategy would involve creating sequences for each table, deviating from our original plan. Conversely, allowing the database to generate IDs simplifies the project, especially considering the ongoing debate between PostgresSQL and MySQL at the time of committing this code.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java (hunk 2 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java]: To adhere to the DRY (Don't Repeat Yourself) principle, I consolidated common attributes into a single class. This decision aligns with our project's current absence of plans for utilizing a microservices architecture.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/repository/BaseRepository.java (hunk 3 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/repository/BaseRepository.java]: This class empowers us to create custom methods not automatically generated by JPA, necessitating the crafting of bespoke queries. Although currently unpopulated, its creation allows us to implement these functionalities without altering existing codebases. This adheres to the second principle of SOLID, namely the open-closed principle.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java (hunk 2 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/entity/BaseEntity.java]: OK.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java (hunk 4 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/base/service/BaseService.java]: This class serves a similar purpose to BaseRepository, and maintaining consistency dictates that we retain both.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/Category.java (hunk 5 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/Category.java]: During the code commit, the database selection was pending. Consequently, I chose to use uppercase letters in table names, following the convention in my workplace, where Oracle DB is utilized. However, post the decision to use PostgreSQL, I acknowledge your point, and I will promptly modify the table names accordingly.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/events/Event.java (hunk 6 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/events/Event.java:24]: By employing the SINGLE_TABLE Inheritance Type, all records consolidate into a single table. This eliminates the necessity of joining two tables, specifically, the ones for offline events and online events. This approach enhances performance, as it allows for seamless retrieval of all events associated with an organizer without the overhead of joining tables.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/TagRepository.java (hunk 8 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/TagRepository.java]: You're correct, and I overlooked this aspect. I appreciate you bringing it to my attention, and I'll be sure to take it into consideration.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/CategoryRepository.java (hunk 9 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/repositories/CategoryRepository.java]: You're absolutely right. As you're aware, I developed the event controllers to assess my work, and this particular method originated from that testing process. Regrettably, I overlooked removing it.

File: src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/enums/Platform.java (hunk 7 above)
Comment (JavaBaz) [reply src/main/java/ir/tehran/jug/projects/eventmanagement/data/entities/enums/Platform.java]: How about these names? 
- **CollaborationService**
- **MeetingApp**
- **OnlineMeetingService** 
- **CommunicationTool**
